x-pack/plugins/security_solution/server/lib/detection_engine/signals/executors/ml.ts

```diff
@@ -0,0 +1,154 @@
+import {
+  AlertInstanceContext,
+  AlertInstanceState,
+  AlertServices,
+} from '../../../../../../alerting/server';
+import { ListClient } from '../../../../../../lists/server';
+import { isJobStarted } from '../../../../../common/machine_learning/helpers';
+import { ExceptionListItemSchema } from '../../../../../common/shared_imports';
+import { SetupPlugins } from '../../../../plugin';
+import { RefreshTypes } from '../../types';
+import { bulkCreateMlSignals } from '../bulk_create_ml_signals';
+import { filterEventsAgainstList } from '../filters/filter_events_against_list';
+import { findMlSignals } from '../find_ml_signals';
+import { BuildRuleMessage } from '../rule_messages';
+import { RuleStatusService } from '../rule_status_service';
+import { MachineLearningRuleAttributes } from '../types';
+import { createErrorsFromShard, createSearchAfterReturnType, mergeReturns } from '../utils';
+
+export const mlExecutor = async ({
+  rule,
+  ml,
+  listClient,
+  exceptionItems,
+  ruleStatusService,
+  services,
+  logger,
+  refresh,
+  buildRuleMessage,
+}: {
+  rule: SavedObject<MachineLearningRuleAttributes>;
+  ml: SetupPlugins['ml'];
+  listClient: ListClient;
+  exceptionItems: ExceptionListItemSchema[];
+  ruleStatusService: RuleStatusService;
+  services: AlertServices<AlertInstanceState, AlertInstanceContext, 'default'>;
+  logger: Logger;
+  refresh: RefreshTypes;
+  buildRuleMessage: BuildRuleMessage;
+}) => {
+  const result = createSearchAfterReturnType();
+  const ruleParams = rule.attributes.params;
+  if (ml == null) {
+    throw new Error('ML plugin unavailable during rule execution');
+  }
+  if (ruleParams.machineLearningJobId == null || ruleParams.anomalyThreshold == null) {
+    throw new Error(
+      [
+        'Machine learning rule is missing job id and/or anomaly threshold:',
+        `job id: "${ruleParams.machineLearningJobId}"`,
+        `anomaly threshold: "${ruleParams.anomalyThreshold}"`,
+      ].join(' ')
+    );
+  }
+
+  // Using fake KibanaRequest as it is needed to satisfy the ML Services API, but can be empty as it is
+  // currently unused by the jobsSummary function.
+  const fakeRequest = {} as KibanaRequest;
+  const summaryJobs = await ml
+    .jobServiceProvider(fakeRequest, services.savedObjectsClient)
+    .jobsSummary([ruleParams.machineLearningJobId]);
+  const jobSummary = summaryJobs.find((job) => job.id === ruleParams.machineLearningJobId);
+
+  if (jobSummary == null || !isJobStarted(jobSummary.jobState, jobSummary.datafeedState)) {
+    const errorMessage = buildRuleMessage(
+      'Machine learning job is not started:',
+      `job id: "${ruleParams.machineLearningJobId}"`,
+      `job status: "${jobSummary?.jobState}"`,
+      `datafeed status: "${jobSummary?.datafeedState}"`
+    );
+    logger.warn(errorMessage);
+    result.warning = true;
+    // TODO: change this to partialFailure since we don't immediately exit rule function and still do actions at the end?
```

Comment:
This is a new TODO: noticed this behavior when adding the warning logic to result. I think this error was added before we created partial errors, would it fit better as a partial error?